### PR TITLE
Move smart utils to its own module

### PR DESCRIPTION
**Pull request by elfosardo**
Smart tools takes a big part of detect_utils.
This patch moves all the smart related tools to its own module and
add a unit test for that module.

Also some refactoring in detect and detect_utils.

### CODE REVIEW

**Comment by elfosardo** [comment]
@dtantsur it passes tox with py2.7 and py3.6, and pep8
Unfortunately there is no test for detect_disks, I can add that here or in a different patch, but I tested it on a baremetal server succesfully:
`
[...]
read_smart: Reading S.M.A.R.T information on /dev/sda
read_smart_scsi: Found S.M.A.R.T information on /dev/sda
read_smart_scsi: Found S.M.A.R.T information on /dev/sda with -d megaraid,0
read_smart_scsi: Found S.M.A.R.T information on /dev/sda with -d megaraid,1
[...]
('disk', 'sda', 'SMART/vendor', 'DELL'),
 ('disk', 'sda', 'SMART/product', 'PERC H710'),
 ('disk', 'sda', 'SMART/serial_number', '0093938205c6ba561f00f2d0d2e03f08'),
 ('disk', 'sda{megaraid,0}', 'SMART/vendor', 'SEAGATE'),
 ('disk', 'sda{megaraid,0}', 'SMART/product', 'ST600MM0006'),
 ('disk', 'sda{megaraid,0}', 'SMART/serial_number', 'S0M3EVL6'),
 ('disk', 'sda{megaraid,0}', 'SMART/health', 'OK'),
 ('disk', 'sda{megaraid,0}', 'SMART/current_drive_temperature', '33'),
 ('disk', 'sda{megaraid,0}', 'SMART/current_drive_temperature_unit', 'C'),
 ('disk', 'sda{megaraid,0}', 'SMART/drive_trip_temperature', '50'),
 ('disk', 'sda{megaraid,0}', 'SMART/drive_trip_temperature_unit', 'C'),
 ('disk', 'sda{megaraid,0}', 'SMART/manufacture_date', 'week 33 of year 2014'),
 ('disk',
  'sda{megaraid,0}',
  'SMART/specified_start_stop_cycle_count_over_lifetime',
  '10000'),
 ('disk', 'sda{megaraid,0}', 'SMART/start_stop_cycle_count', '310'),
 ('disk',
  'sda{megaraid,0}',
  'SMART/specified_load_count_over_lifetime',
  '300000'),
 ('disk', 'sda{megaraid,0}', 'SMART/load_count', '10909'),
 ('disk', 'sda{megaraid,0}', 'SMART/blocks_sent', '3525471124'),
 ('disk', 'sda{megaraid,0}', 'SMART/blocks_received', '437590566'),
 ('disk', 'sda{megaraid,0}', 'SMART/blocks_read_from_cache', '16227520'),
 ('disk', 'sda{megaraid,0}', 'SMART/power_on_hours', '12544.85'),
 ('disk', 'sda{megaraid,0}', 'SMART/read_total_corrected_errors', '2403625015'),
 ('disk', 'sda{megaraid,0}', 'SMART/read_gigabytes_processed', '4059.874'),
 ('disk', 'sda{megaraid,0}', 'SMART/read_total_uncorrected_errors', '0'),
 ('disk', 'sda{megaraid,0}', 'SMART/write_total_corrected_errors', '0'),
 ('disk', 'sda{megaraid,0}', 'SMART/write_gigabytes_processed', '8458.141'),
 ('disk', 'sda{megaraid,0}', 'SMART/write_total_uncorrected_errors', '0'),
 ('disk',
  'sda{megaraid,0}',
  'SMART/verify_total_corrected_errors',
  '2521365608'),
 ('disk', 'sda{megaraid,0}', 'SMART/verify_gigabytes_processed', '51750.086'),
 ('disk', 'sda{megaraid,0}', 'SMART/verify_total_uncorrected_errors', '0'),
 ('disk', 'sda{megaraid,0}', 'SMART/non_medium_errors_count', '223'),
 ('disk', 'sda{megaraid,1}', 'SMART/vendor', 'SEAGATE'),
 ('disk', 'sda{megaraid,1}', 'SMART/product', 'ST600MM0006'),
 ('disk', 'sda{megaraid,1}', 'SMART/serial_number', 'S0M3EY2Z'),
 ('disk', 'sda{megaraid,1}', 'SMART/health', 'OK'),
 ('disk', 'sda{megaraid,1}', 'SMART/current_drive_temperature', '32'),
 ('disk', 'sda{megaraid,1}', 'SMART/current_drive_temperature_unit', 'C'),
 ('disk', 'sda{megaraid,1}', 'SMART/drive_trip_temperature', '50'),
 ('disk', 'sda{megaraid,1}', 'SMART/drive_trip_temperature_unit', 'C'),
 ('disk', 'sda{megaraid,1}', 'SMART/manufacture_date', 'week 33 of year 2014'),
 ('disk',
  'sda{megaraid,1}',
  'SMART/specified_start_stop_cycle_count_over_lifetime',
  '10000'),
 ('disk', 'sda{megaraid,1}', 'SMART/start_stop_cycle_count', '310'),
 ('disk',
  'sda{megaraid,1}',
  'SMART/specified_load_count_over_lifetime',
  '300000'),
 ('disk', 'sda{megaraid,1}', 'SMART/load_count', '11025'),
 ('disk', 'sda{megaraid,1}', 'SMART/blocks_sent', '3522926458'),
 ('disk', 'sda{megaraid,1}', 'SMART/blocks_received', '475554044'),
 ('disk', 'sda{megaraid,1}', 'SMART/blocks_read_from_cache', '35568987'),
 ('disk', 'sda{megaraid,1}', 'SMART/power_on_hours', '12551.22'),
 ('disk', 'sda{megaraid,1}', 'SMART/read_total_corrected_errors', '2397327258'),
 ('disk', 'sda{megaraid,1}', 'SMART/read_gigabytes_processed', '4332.157'),
 ('disk', 'sda{megaraid,1}', 'SMART/read_total_uncorrected_errors', '0'),
 ('disk', 'sda{megaraid,1}', 'SMART/write_total_corrected_errors', '0'),
 ('disk', 'sda{megaraid,1}', 'SMART/write_gigabytes_processed', '8353.223'),
 ('disk', 'sda{megaraid,1}', 'SMART/write_total_uncorrected_errors', '0'),
 ('disk',
  'sda{megaraid,1}',
  'SMART/verify_total_corrected_errors',
  '1924628922'),
 ('disk', 'sda{megaraid,1}', 'SMART/verify_gigabytes_processed', '51729.903'),
 ('disk', 'sda{megaraid,1}', 'SMART/verify_total_uncorrected_errors', '0'),
 ('disk', 'sda{megaraid,1}', 'SMART/non_medium_errors_count', '266'),

`

**Comment by tbreeds** [comment]
Actually this fails unit tests locally:

```
hardware.tests.test_smart_utils.TestSmartUtils.test_read_smart_call_smart_ata
-----------------------------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    b'Traceback (most recent call last):'
    b'  File "/home/tony/projects/OSF/related-projects/hardware/.tox/cover/lib/python3.7/site-packages/mock/mock.py", line 1330, in patched'
    b'    return func(*args, **keywargs)'
    b'  File "/home/tony/projects/OSF/related-projects/hardware/hardware/tests/test_smart_utils.py", line 110, in test_read_smart_call_smart_ata'
    b'    mock_ata.assert_called()'
    b'  File "/home/tony/projects/OSF/related-projects/hardware/.tox/cover/lib/python3.7/site-packages/mock/mock.py", line 908, in assert_called'
    b'    raise AssertionError(msg)'
    b"AssertionError: Expected 'read_smart_ata' to have been called."
    b''


======
Totals
======
Ran: 175 tests in 0.7085 sec.
 - Passed: 172
 - Skipped: 1
 - Expected Fail: 0
 - Unexpected Success: 0
 - Failed: 2
Sum of execute time for each test: 2.2295 sec.

==============
Worker Balance
==============
 - Worker 0 (44 tests) => 0:00:00.512590
 - Worker 1 (44 tests) => 0:00:00.663453
 - Worker 2 (44 tests) => 0:00:00.568319
 - Worker 3 (43 tests) => 0:00:00.594153
ERROR: InvocationError for command '/home/tony/projects/OSF/related-projects/hardware/.tox/cover/bin/stestr run' (exited with code 1)
___________________________________ summary ____________________________________
ERROR:   py35: commands failed
ERROR:   py36: commands failed
ERROR:   py37: commands failed
ERROR:   py27: commands failed
  pep8: commands succeeded
  pylint: commands succeeded
ERROR:   cover: commands failed
```

**Comment by elfosardo** [comment]
@tbreeds that's really weird, I ran all the tests again and I can't reproduce the failure locally:
```
tox -r -vv -epy27

======
Totals
======
Ran: 175 tests in 0.0000 sec.
 - Passed: 174
 - Skipped: 1
 - Expected Fail: 0
 - Unexpected Success: 0
 - Failed: 0
Sum of execute time for each test: 1.7326 sec.

==============
Worker Balance
==============
 - Worker 0 (1 tests) => 0:00:00.139432
 - Worker 1 (23 tests) => 0:00:00.188522
 - Worker 2 (24 tests) => 0:00:00.218121
 - Worker 3 (23 tests) => 0:00:00.271608
 - Worker 4 (24 tests) => 0:00:00.211324
 - Worker 5 (25 tests) => 0:00:00.248170
 - Worker 6 (25 tests) => 0:00:00.222089
 - Worker 7 (30 tests) => 0:00:00.306774
py27 finish: runtests after 2.49 seconds
py27 start: run-test-post 
py27 finish: run-test-post after 0.00 seconds

#################

tox -r -vv -epy36

======
Totals
======
Ran: 175 tests in 0.5888 sec.
 - Passed: 174
 - Skipped: 1
 - Expected Fail: 0
 - Unexpected Success: 0
 - Failed: 0
Sum of execute time for each test: 1.9013 sec.

==============
Worker Balance
==============
 - Worker 0 (1 tests) => 0:00:00.197812
 - Worker 1 (25 tests) => 0:00:00.289880
 - Worker 2 (25 tests) => 0:00:00.173787
 - Worker 3 (24 tests) => 0:00:00.238555
 - Worker 4 (26 tests) => 0:00:00.316019
 - Worker 5 (24 tests) => 0:00:00.350081
 - Worker 6 (25 tests) => 0:00:00.257513
 - Worker 7 (25 tests) => 0:00:00.165627

#################

tox -r -vv -epy37

======
Totals
======
Ran: 175 tests in 0.5939 sec.
 - Passed: 174
 - Skipped: 1
 - Expected Fail: 0
 - Unexpected Success: 0
 - Failed: 0
Sum of execute time for each test: 2.1380 sec.

==============
Worker Balance
==============
 - Worker 0 (1 tests) => 0:00:00.220363
 - Worker 1 (1 tests) => 0:00:00.169609
 - Worker 2 (12 tests) => 0:00:00.194171
 - Worker 3 (31 tests) => 0:00:00.351530
 - Worker 4 (31 tests) => 0:00:00.265367
 - Worker 5 (31 tests) => 0:00:00.297737
 - Worker 6 (35 tests) => 0:00:00.456168
 - Worker 7 (33 tests) => 0:00:00.309981
py37 finish: runtests after 2.79 seconds
py37 start: run-test-post 
py37 finish: run-test-post after 0.00 seconds

##########

tox -r -vv

  py35: commands succeeded
  py36: commands succeeded
  py37: commands succeeded
  py27: commands succeeded
  pep8: commands succeeded
  pylint: commands succeeded
  cover: commands succeeded
  congratulations :)
```

**Comment by ErwanAliasr1** [comment]
Heya, any updates on this topic ?
I'll have to make some updates on the SMART code

**Comment by elfosardo** [comment]
@ErwanAliasr1 , this change has been already approved by @tbreeds and @dtantsur , best way to speed things up would be leaving a review and help merging the change :)

File: hardware/smart_utils.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+                              values[title]))
+            continue
+
+        except Exception:
+            sys.stderr.write("read_smart: failed to read line : %s\n" % line)
+            continue
+
+
+def read_smart(hwlst, device, optional_flag=""):
+    if not which("smartctl"):
```
Comment (tbreeds) [inline hardware/smart_utils.py]: The reason this is failing locally is I don't have a smartctl so this 'which' call is returning None.  and therefore the rest of the test isn;t run resulting in the fail.

Please also mock this out in test_smart_utils.py

**Comment by tbreeds** [comment]
So I worked out why it's failing for me and not you.  Please fix that and I think we're good to go

File: hardware/smart_utils.py (hunk 1 above)
Comment (elfosardo) [reply hardware/smart_utils.py]: done

**Comment by elfosardo** [comment]
@tbreeds thanks, I've added the mock and ran the tests again (removing smartctl first from the path), and it seems ok

**Pull request merged**